selfdrive/common/tests/test_util.cc

```diff
@@ -0,0 +1,98 @@
+
+#include <dirent.h>
+#include <sys/types.h>
+
+#include <algorithm>
+#include <climits>
+#include <random>
+#include <string>
+
+#define CATCH_CONFIG_MAIN
+#include "catch2/catch.hpp"
+#include "selfdrive/common/util.h"
+
+std::string random_string(size_t length) {
+  auto randchar = []() -> char {
+    const char charset[] =
+        "0123456789"
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        "abcdefghijklmnopqrstuvwxyz";
+    const size_t max_index = (sizeof(charset) - 1);
+    return charset[rand() % max_index];
+  };
+  std::string str(length, 0);
+  std::generate_n(str.begin(), length, randchar);
+  return str;
+}
+
+template <typename OutputIt>
+void random_bytes(OutputIt first, OutputIt last) {
+  std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned char> rbe;
+  std::generate(first, last, std::ref(rbe));
+}
+
+TEST_CASE("util::read_file") {
+  auto read_file = [](const std::string &fn) -> std::string {
+    std::ifstream f(fn);
+    std::stringstream buffer;
+    buffer << f.rdbuf();
+    return buffer.str();
+  };
+
+  SECTION("read /proc") {
+    DIR *d = opendir("/proc");
+    struct dirent *de = nullptr;
+    while ((de = readdir(d))) {
+      const std::string path = util::string_format("/proc/%s/cmdline", de->d_name);
+      std::string ret1 = util::read_file(path);
+      std::string ret2 = read_file(path);
+      REQUIRE(ret1 == ret2);
+    }
+    closedir(d);
+  }
+  SECTION("read ascii file") {
+    std::vector<std::string> test_data;
```

Comment:
Why not one long string?

Reply:
I used a single temp file to continue writing to it.

Reply:
What's wrong with an std::string of 64k?